File: 3754-maximum-manhattan-distance-after-k-changes/maximum-manhattan-distance-after-k-changes.py
class Solution:
    def maxDistance(self, s: str, k: int) -> int:
        lst = [{"N", "E"}, {"N", "W"}, {"S", "E"}, {"S", "W"}]
        ans = 0

        for diag in lst:
            kk, dist = k, 0
            for ch in s:
                if ch in diag or kk:
                    dist += 1
                    kk -= ch not in diag
                else:
                    dist -= 1
                ans = max(ans, dist)

        return ans

File: 3754-maximum-manhattan-distance-after-k-changes/maximum-manhattan-distance-after-k-changes.cpp
class Solution {
public:
    int maxDistance(string s, int k) {
        int n = s.size(), ans = 1, hor = 0, ver = 0;

        for (int i = 0; i < n; ++i) {
            ver += s[i] == 'N' ? 1 : s[i] == 'S' ? -1 : 0;
            hor += s[i] == 'W' ? 1 : s[i] == 'E' ? -1 : 0;
            ans = max(ans, min(abs(hor) + abs(ver) + 2 * k, i + 1));
        }
        
        return ans;
    }
};